File: secator/tasks/urlfinder.py
import validators
from collections import defaultdict
from urllib.parse import urlparse, urlunparse, parse_qs

from secator.definitions import HOST, URL, DELAY, DEPTH, FILTER_CODES, FILTER_REGEX, FILTER_SIZE, FILTER_WORDS, MATCH_CODES, MATCH_REGEX, MATCH_SIZE, MATCH_WORDS, FOLLOW_REDIRECT, METHOD, PROXY, RATE_LIMIT, RETRIES, THREADS, TIMEOUT, USER_AGENT, HEADER, OPT_NOT_SUPPORTED  # noqa: E501
from secator.output_types import Url
from secator.decorators import task
from secator.serializers import JSONSerializer
from secator.tasks._categories import HttpCrawler


URLFINDER_SOURCES = [
	'alienvault',
	'commoncrawl',
	'urlscan',
	'waybackarchive',
	'virustotal'
]


@task()
class urlfinder(HttpCrawler):
	"""Find URLs in text."""
	cmd = 'urlfinder'
	input_types = [HOST, URL]
	output_types = [Url]
	item_loaders = [JSONSerializer()]
	json_flag = '-j'
	tags = ['pattern', 'scan']
	file_flag = '-list'
	input_flag = '-d'
	version_flag = '-version'
	opts = {
		'sources': {'type': str, 'help': f'Sources to use (comma-delimited) ({", ".join(URLFINDER_SOURCES)})', 'required': False},  # noqa: E501
		'exclude_sources': {'type': str, 'help': 'Sources to exclude (comma-delimited)', 'required': False},
		'version': {'type': bool, 'help': 'Version', 'required': False},
		'stats': {'type': bool, 'help': 'Display source statistics', 'required': False},
		'max_param_occurrences': {'type': int, 'help': 'Max occurrences for the same parameter in the same URL before discarding next results', 'required': False, 'default': 10, 'internal': True},  # noqa: E501
	}
	opt_key_map = {
		'sources': 's',
		'exclude_sources': 'es',
		'url_scope': 'us',
		'url_out_scope': 'uos',
		'field_scope': 'fs',
		'no_scope': 'ns',
		'display_out_scope': 'do',
		'match': 'm',
		'filter': 'f',
		HEADER: OPT_NOT_SUPPORTED,
		DELAY: OPT_NOT_SUPPORTED,
		DEPTH: OPT_NOT_SUPPORTED,
		FILTER_CODES: OPT_NOT_SUPPORTED,
		FILTER_REGEX: OPT_NOT_SUPPORTED,
		FILTER_SIZE: OPT_NOT_SUPPORTED,
		FILTER_WORDS: OPT_NOT_SUPPORTED,
		MATCH_CODES: OPT_NOT_SUPPORTED,
		MATCH_REGEX: OPT_NOT_SUPPORTED,
		MATCH_SIZE: OPT_NOT_SUPPORTED,
		MATCH_WORDS: OPT_NOT_SUPPORTED,
		FOLLOW_REDIRECT: OPT_NOT_SUPPORTED,
		METHOD: OPT_NOT_SUPPORTED,
		PROXY: 'proxy',
		RATE_LIMIT: 'rl',
		RETRIES: OPT_NOT_SUPPORTED,
		THREADS: OPT_NOT_SUPPORTED,
		TIMEOUT: 'timeout',
		USER_AGENT: OPT_NOT_SUPPORTED,
	}
	encoding = 'ansi'
	install_version = 'v0.0.3'
	install_cmd = 'go install -v github.com/projectdiscovery/urlfinder/cmd/urlfinder@[install_version]'
	github_handle = 'projectdiscovery/urlfinder'
	proxychains = False
	proxy_socks5 = True
	proxy_http = True
	profile = 'io'

	@staticmethod
	def before_init(self):
		# Call parent's before_init to process raw HTTP request
		HttpCrawler.before_init(self)

		for idx, input in enumerate(self.inputs):
			if validators.url(input):
				self.inputs[idx] = urlparse(input).netloc

	@staticmethod
	def on_init(self):
		self.max_param_occurrences = self.get_opt_value('max_param_occurrences')
		self.seen_params = defaultdict(lambda: defaultdict(int))

	@staticmethod
	def on_json_loaded(self, item):
		parsed_url = urlparse(item['url'])
		base_url = urlunparse(parsed_url._replace(query="", fragment=""))  # Remove query & fragment
		query_params = parse_qs(parsed_url.query)
		current_params = set(query_params.keys())
		for param in current_params:
			self.seen_params[base_url][param] += 1
			if self.seen_params[base_url][param] > int(self.max_param_occurrences):
				return
		yield Url(url=item['url'], host=parsed_url.hostname, extra_data={'source': item['source']})

File: rust/crates/secator-tasks/src/urlfinder.rs
//! urlfinder — passive URL discovery (Python `secator/tasks/urlfinder.py`).
//!
//! Same shape as `xurlfind3r`: URL inputs are reduced to netloc in
//! `before_init`, and `on_json_loaded` emits `Url(tags=["passive"])` with a
//! per-(base_url, param) cap (Python `max_param_occurrences`, default 10) to
//! prevent flooding when many query-string variants exist.

use std::collections::BTreeMap;

use secator_model::{Map, OutputItem, Url};
use secator_options::{FileMode, InputWiring, KeyMap, OptSchema, OptSpec, OptType, SingleMode};
use secator_runner::{empty_output_maps, HookCtx, HookRegistry, ItemLoader, ValidatorRegistry};
use serde_json::Value;
use url::Url as UrlParser;

use crate::TaskSpec;

pub static SPEC: TaskSpec = TaskSpec {
    name: "urlfinder",
    description: "Find URLs in passive sources (alienvault, urlscan, wayback, …).",
    cmd: "urlfinder",
    input_types: &["host", "url"],
    output_types: &["url"],
    tags: &["url", "crawl", "passive"],
    json_flag: Some("-j"),
    input_wiring: InputWiring { single: SingleMode::Flag("-d"), file: FileMode::Flag("-list") },
    item_loaders: &[ItemLoader::Json],
    input_chunk_size: 1,
    on_json_loaded: Some(on_json_loaded),
    on_regex_loaded: None,
    output_maps: empty_output_maps,
    discriminator: None,
    hooks: HOOKS,
    validators: ValidatorRegistry::EMPTY,
    schema: build_schema,
    install: secator_runner::InstallSpec {
        version: Some("v0.0.3"),
        cmd: Some("go install -v github.com/projectdiscovery/urlfinder/cmd/urlfinder@[install_version]"),
        github_handle: Some("projectdiscovery/urlfinder"),
        ..secator_runner::InstallSpec::EMPTY
    },
    proxy_caps: secator_runner::ProxyCaps { proxychains: false, proxy_http: true, proxy_socks5: false },
    encoding: "utf-8",
    ignore_return_code: false,
    requires_sudo: false,
};

static HOOKS: HookRegistry = HookRegistry {
    before_init: &[before_init_normalize_inputs],
    ..HookRegistry::EMPTY
};

fn before_init_normalize_inputs(
    ctx: &mut HookCtx,
    runner: &mut secator_runner::CommandRunner,
) {
    for i in runner.inputs.iter_mut() {
        if let Ok(u) = UrlParser::parse(i) {
            if let Some(host) = u.host_str() {
                *i = host.to_string();
            }
        }
    }
    let cap = runner
        .opts
        .get("max_param_occurrences")
        .and_then(|v| v.parse::<u32>().ok())
        .unwrap_or(10);
    ctx.state.insert("urlfinder:cap".into(), cap.to_string());
}

fn on_json_loaded(ctx: &mut HookCtx, item: Map) -> Vec<OutputItem> {
    let url_str = match item.get("url").and_then(|v| v.as_str()) {
        Some(s) => s.to_string(),
        None => return Vec::new(),
    };
    let cap: u32 = ctx
        .state
        .get("urlfinder:cap")
        .and_then(|s| s.parse().ok())
        .unwrap_or(10);
    let parsed = match UrlParser::parse(&url_str) {
        Ok(u) => u,
        Err(_) => return Vec::new(),
    };
    let base_url = {
        let mut b = parsed.clone();
        b.set_query(None);
        b.set_fragment(None);
        b.to_string()
    };
    let host = parsed.host_str().unwrap_or("").to_string();
    for (k, _) in parsed.query_pairs() {
        let key = format!("urlfinder:p:{base_url}:{}", k.as_ref());
        let count: u32 = ctx
            .state
            .get(&key)
            .and_then(|s| s.parse().ok())
            .unwrap_or(0)
            + 1;
        ctx.state.insert(key, count.to_string());
        if count > cap {
            return Vec::new();
        }
    }
    let mut extra: BTreeMap<String, Value> = BTreeMap::new();
    if let Some(s) = item.get("source").and_then(|v| v.as_str()) {
        extra.insert("source".into(), Value::String(s.to_string()));
    }
    let extra: Map = extra.into_iter().collect();
    vec![OutputItem::Url(Url {
        url: url_str,
        host,
        extra_data: extra,
        tags: vec!["passive".into()],
        ..Default::default()
    })]
}

fn build_schema() -> OptSchema {
    let mut s = OptSchema { opt_prefix: "-", ..OptSchema::default() };
    s.meta_opts = crate::meta_opts::opts_http_base();
    crate::meta_opts::apply_config_defaults(&mut s.meta_opts);
    // Python `opt_key_map` shortens common flags.
    s.key_map.insert("proxy".into(), KeyMap::Flag("proxy".into()));
    s.key_map.insert("rate_limit".into(), KeyMap::Flag("rl".into()));
    s.key_map.insert("timeout".into(), KeyMap::Flag("timeout".into()));
    s.key_map.insert("sources".into(), KeyMap::Flag("s".into()));
    s.key_map.insert("exclude_sources".into(), KeyMap::Flag("es".into()));
    for k in [
        "header", "delay", "depth", "filter_codes", "filter_regex",
        "filter_size", "filter_words", "match_codes", "match_regex",
        "match_size", "match_words", "follow_redirect", "method",
        "retries", "threads", "user_agent",
    ] {
        s.key_map.insert(k.into(), KeyMap::NotSupported);
    }
    s.opts = vec![
        OptSpec {
            name: "sources",
            ty: OptType::Str,
            short: None,
            is_flag: false,
            default: None,
            help: "Comma-delimited sources (alienvault, commoncrawl, urlscan, wayback, virustotal)",
            internal: false,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
        OptSpec {
            name: "exclude_sources",
            ty: OptType::Str,
            short: None,
            is_flag: false,
            default: None,
            help: "Comma-delimited sources to exclude",
            internal: false,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
        OptSpec {
            name: "stats",
            ty: OptType::Bool,
            short: None,
            is_flag: true,
            default: None,
            help: "Display source statistics",
            internal: false,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
        OptSpec {
            name: "max_param_occurrences",
            ty: OptType::Int,
            short: None,
            is_flag: false,
            default: Some("10"),
            help: "Cap per (base_url, param) before dropping further dups",
            internal: true,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
    ];
    s
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn map_from(v: Value) -> Map {
        v.as_object().cloned().unwrap()
    }

    #[test]
    fn emits_url_with_passive_tag() {
        let mut ctx = HookCtx::default();
        ctx.state.insert("urlfinder:cap".into(), "10".into());
        let out = on_json_loaded(
            &mut ctx,
            map_from(json!({"url": "https://example.com/a", "source": "wayback"})),
        );
        if let Some(OutputItem::Url(u)) = out.first() {
            assert_eq!(u.host, "example.com");
            assert!(u.tags.iter().any(|t| t == "passive"));
        } else {
            panic!("expected Url");
        }
    }
}
